1.0.0 Migration Issues

Collection of issues in migrations:

#4 (prov:Agent set to Email instead of software agent, potential SHACL problem

```
366f655c2421a873c4bc6d37c8a0c1d2f04a2ce |      12329 | cyril.matthey-doret/sequence-loopability
Caused by: ammonite.ops.ShelloutException: CommandResult 1;
Error: Validation Report;
Conforms: False;
Results (2):;
Constraint Violation in ClassConstraintComponent (http://www.w3.org/ns/shacl#ClassConstraintComponent):;
Severity: sh:Violation;
Source Shape: [ sh:class prov:SoftwareAgent ;
sh:maxCount Literal("1", datatype=xsd:integer) ;
sh:minCount Literal("1", datatype=xsd:integer) ;
sh:path prov:agent ];
Focus Node: <https:
//david.dev.renku.ch/activities/bc7ddb08d42a4cbe8caa2815f4e86264/association>;
Value Node: <mailto:cyril.matthey-doret@pasteur.fr>;
Result Path: prov:agent;
Message: Value does not have class prov:SoftwareAgent;
Constraint Violation in ClassConstraintComponent (http://www.w3.org/ns/shacl#ClassConstraintComponent):;
Severity: sh:Violation;
Source Shape: [ sh:class prov:SoftwareAgent ;
sh:maxCount Literal("1", datatype=xsd:integer) ;
sh:minCount Literal("1", datatype=xsd:integer) ;
sh:path prov:agent ];
Focus Node: <https://david.dev.renku.ch/activities/4b3729f7d3704543b12c859e1b95a709/association>;
Value Node: <mailto:cyril.matthey-doret@pasteur.fr>;
Result Path: prov:agent;
Message: Value does not have class
 prov:SoftwareAgent;
 ;
 Couldn't export: Invalid Knowledge Graph data;
 ;
```



#8 Submodule Migration issue
```
cafeb1cbed77fdc312f4c3af278bbb03ba068086 |        145 | erbou/celegans-genome-analysis-1


Applying migration m_0003__0_pyld2...;
Applying migration m_0003__1_jsonld...;
Applying migration m_0003__2_initial...;
Applying migration m_0004__0_pyld2...;
Applying migration m_0004__submodules...;
Applying migration m_0003__0_pyld2...;
Applying migration m_0003__1_jsonld...;
Applying migration m_0003__2_initial...;
Applying migration m_0004__0_pyld2...;
Applying migration m_0004__submodules...;
Applying migration m_0005__1_pyld2...;
Applying migration m_0005__2_cwl...;
Caching commit 1\rCaching commit 2\rCaching commit 3\rCaching commit 4\rCaching commit 5\rCaching commit 6\rCaching commit 7\rCaching commit 8\rCaching commit 9\rCaching commit 10\rCaching commit 11\rCaching commit 12\rCaching commit 13\rCaching commit 14\rCaching commit 15\rCaching commit 16\rCaching commit 17\r                                        \rProcessing commit 1/1\rApplying migration m_0006__dataset_context...;
Applying migration m_0007__source_url...;
Applying migration m_0008__dataset_metadata...;
Applying migration m_0009__new_metadata_storage...;
Processing commits 1/3 e0cf627c320eaaa9e98746ea810acfd9e71c5947\rProcessing commits 2/3 e6c764abbdbfb72247fd858e33c4263957ef930a;
Warning: Cannot process commit 'e6c764abbdbfb72247fd858e33c4263957ef930a' - Exception: Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0009__new_metadata_storage.py", line 194, in generate_new_metadata;
_process_workflows(client=client, activity_gateway=activity_gateway, commit=commit, remove=remove);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0009__new_metadata_storage.py", line 336, in _process_workflows;
workflow = old_schema.Activity.from_yaml(path=path, client=client);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/models/v9.py", line 807, in from_yaml;
data = jsonld.read_yaml(path);
File "/usr/lib/python3.8/site-packages/renku/core/models/jsonld.py", line 62, in read_yaml;
with Path(path).open(mode="r") as fp:;
File "/usr/lib/python3.8/pathlib.py", line 1222, in open;
return io.open(self, mode, buffering, encoding, errors, newline,;
File "/usr/lib/python3.8/pathlib.py", line 1078,in _opener;
return self._accessor.open(self, flags, mode);
FileNotFoundError: [Errno 2] No such file or directory: '.renku/workflow/4c834c484234422a8dc768070b94f9b6_code_runSTAR.sh.yaml';
;
Processing commits 3/3 1e233fbda43e328750b680bb3c69dbe9d5a3833c\rSuccessfully applied 11 migrations.;
Applying migration m_0003__0_pyld2...;
Applying migration m_0003__1_jsonld...;
Applying migration m_0003__2_initial...;
Applying migration m_0004__0_pyld2...;
Applying migration m_0004__submodules...;
Applying migration m_0003__0_pyld2...;
Applying migration m_0003__1_jsonld...;
Applying migration m_0003__2_initial...;
Applying migration m_0004__0_pyld2...;
Applying migration m_0004__submodules...;
Applying migration m_0005__1_pyld2...;
Applying migration m_0005__2_cwl...;
Caching commit 1\rCaching commit 2\rCaching commit 3\rCaching commit 4\rCaching commit 5\rCaching commit 6\rCaching commit 7\rCaching commit 8\rCaching commit 9\rCaching commit 10\rCaching commit 11\rCaching commit 12\rCaching commit 13\rCaching commit 14\rCaching commit 15\r\rProcessing commit 1/1\rApplying migration m_0006__dataset_context...;
Applying migration m_0007__source_url...;
Applying migration m_0008__dataset_metadata...;
Applying migration m_0009__new_metadata_storage...;
Processing commits 1/3 826343464f7de81f2bee38c1332cd217b014e00f\rProcessing commits 2/3 6fad8b741b5cca653c3fc8535ff1fc1015c78808;
Warning: Cannot process commit '6fad8b741b5cca653c3fc8535ff1fc1015c78808' - Exception: Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0009__new_metadata_storage.py", line 194, in generate_new_metadata;
_process_workflows(client=client, activity_gateway=activity_gateway, commit=commit, remove=remove);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0009__new_metadata_storage.py", line 336, in _process_workflows;
workflow = old_schema.Activity.from_yaml(path=path, client=client);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/models/v9.py", line 807, in from_yaml;
data = jsonld.read_yaml(path);
File "/usr/lib/python3.8/site-packages/renku/core/models/jsonld.py", line 62, in read_yaml;
with Path(path).open(mode="r") as fp:;
File "/usr/lib/python3.8/pathlib.py", line 1222, in open;
return io.open(self, mode, buffering, encoding, errors, newline,;
File "/usr/lib/python3.8/pathlib.py", line 1078, in _opener;
return self._accessor.open(self, flags, mode);
FileNotFoundError: [Errno 2] No such file or directory: '.renku/workflow/7137679420a54b7b868ecf2d60ddef9a_code_runSTAR.sh.yaml';
;
Processing commits 3/3 ee9a022644e571da8d75d49e398cc3df103a04a0\rSuccessfully applied 11 migrations.;
Error: Couldn't execute migration;
;
Traceback (most recent call last):;
File "/usr/lib/python3.8/shutil.py", line 791, in move;
os.rename(src, real_dst);
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/celegans-genome-analysis-1--1300193730710223129/.renku/vendors/renkulab.io/gitlab/erbou/celegans-genome-mapping-3/.renku/vendors/renkulab.io/gitlab/erbou/celegans-genome/data/celegans/metadata.yml' -> '/tmp/celegans-genome-analysis-1--1300193730710223129/.renku/vendors/renkulab.io/gitlab/erbou/celegans-genome-mapping-3/data/wbcel235/celegans-genome/metadata.yml';
;
During handling of the above exception, another exception occurred:;
;
Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0004__submodules.py", line 129, in _migrate_submodule_based_datasets;
shutil.move(target, path);
File "/usr/lib/python3.8/shutil.py", line 811, in move;
copy_function(src, real_dst);
File "/usr/lib/python3.8/shutil.py", line 435, in copy2;
copyfile(src, dst, follow_symlinks=follow_symlinks);
File "/usr/lib/python3.8/shutil.py", line 264, in copyfile;
with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:;
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/celegans-genome-analysis-1--1300193730710223129/.renku/vendors/renkulab.io/gitlab/erbou/celegans-genome-mapping-3/.renku/vendors/renkulab.io/gitlab/erbou/celegans-genome/data/celegans/metadata.yml';
;
During handling of the above exception, another exception occurred:;
;
Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/core/management/migrate.py", line 157, in migrate;
module.migrate(migration_context);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0004__submodules.py", line 43, in migrate;
_migrate_submodule_based_datasets(migration_context.client);
File "/usr/lib/python3.8/site-packages/inject/__init__.py", line 342, in injection_wrapper;
return sync_func(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0004__submodules.py", line 131, in _migrate_submodule_based_datasets;
raise errors.InvalidFileOperation(f"File was not found: {target}");
renku.core.errors.InvalidFileOperation: File was not found: /tmp/celegans-genome-analysis-1--1300193730710223129/.renku/vendors/renkulab.io/gitlab/erbou/celegans-genome-mapping-3/.renku/vendors/renkulab.io/gitlab/erbou/celegans-genome/data/celegans/metadata.yml;
;
The above exception was the direct cause of the following exception:;
;
Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/core/management/migrate.py", line 157, in migrate;
module.migrate(migration_context);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0004__submodules.py", line 43, in migrate;
_migrate_submodule_based_datasets(migration_context.client);
File "/usr/lib/python3.8/site-packages/inject/__init__.py", line 342, in injection_wrapper;
return sync_func(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0004__submodules.py", line 91, in _migrate_submodule_based_datasets;
migrate(skip_template_update=True, skip_docker_update=True);
File "/usr/lib/python3.8/site-packages/inject/__init__.py", line 342, in injection_wrapper;
return sync_func(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrate.py", line 159, in migrate;
raise MigrationError("Couldn't execute migration") from e;
renku.core.errors.MigrationError: Couldn't execute migration;
;
The above exception was the direct cause of the following exception:;
;
Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/cli/exception_handler.py", line 89, in main;
return super().main(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/click/core.py", line 1062, in main;
rv = self.invoke(ctx);
File "/usr/lib/python3.8/site-packages/click/core.py", line 1668,in invoke;
return _process_result(sub_ctx.command.invoke(sub_ctx));
File "/usr/lib/python3.8/site-packages/click/core.py", line 1404, in invoke;
return ctx.invoke(self.callback, **ctx.params);
File "/usr/lib/python3.8/site-packages/click/core.py", line 763, in invoke;
return __callback(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/cli/migrate.py", line 109, in migrate;
result = command.build().execute(;
File "/usr/lib/python3.8/site-packages/renku/core/management/command_builder/command.py", line 256, in execute;
hook(self, context, result, *args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/command_builder/command.py", line 195, in _post_hook;
raise result.error;File "/usr/lib/python3.8/site-packages/renku/core/management/command_builder/command.py", line 242, in execute;
output = context["click_context"].invoke(self._operation, *args, **kwargs);
File "/usr/lib/python3.8/site-packages/click/core.py", line 763, in invoke;
return __callback(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/commands/migrate.py", line 164, in _migrate_project;
return migrate(;
File "/usr/lib/python3.8/site-packages/inject/__init__.py", line 342, in injection_wrapper;
return sync_func(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrate.py", line 159, in migrate;
raise MigrationError("Couldn't execute migration") from e;
renku.core.errors.MigrationError: Couldn't execute migration;

```



#9 SHACL Error, License is an IRI
```
967f1bc6ae49693bc730496918f4863eb7c67274 |       1872 | ren.ku/dataverse-import

Error: Validation Report;
Conforms: False;
Results (1):;
Constraint Violation in NodeKindConstraintComponent (http://www.w3.org/ns/shacl#NodeKindConstraintComponent):;
Severity: sh:Violation;
Source Shape: [ sh:nodeKind sh:Literal ;
sh:or ( [ sh:datatype xsd:string ;
sh:nodeKind sh:Literal ] [ sh:nodeKind sh:BlankNodeOrIRI ] ) ;
sh:path schema:license ];
Focus Node: <https://david.dev.renku.ch/datasets/ee318fea6df64641a08c7e3cb0350fff>;
Value Node: <https://creativecommons.org/publicdomain/zero/1.0/>;
Result Path: schema:license;
Message: Value is not of Node Kind sh:Literal;
;
Couldn't export: Invalid Knowledge Graph data;
```


#12 SHACL needs `schema:image` on Dataset node
```
Node <https://david.dev.renku.ch/datasets/617f2eaf232917bb0b046893e3a043e7> is closed. It cannot have value: <https://david.dev.renku.ch/datasets/652eb758-d7d4-4a7f-b76e-6157c21adb20/images/0>

617f2eaf232917bb0b04682549b9230fc337db5b |      13997 | pamela.delgado/forked-test-streamlit-new-base-image

 Error: Validation Report;
 Conforms: False;
 Results (2):;
 Constraint Violation in ClosedConstraintComponent (http://www.w3.org/ns/shacl#ClosedConstraintComponent):;
 Severity: sh:Violation;
 Source Shape: [ rdf:type sh:NodeShape ;
 sh:closed Literal("true" = True, datatype=xsd:boolean) ;
 sh:ignoredProperties ( rdf:type ) ;
 sh:property [ sh:class schema:DigitalDocument ;
 sh:path schema:hasPart ], [ sh:class schema:Language ;
 sh:path schema:inLanguage ], [ sh:class schema:Person ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:path schema:creator ], [ sh:class schema:PublicationEvent ;
 sh:path schema:subjectOf ], [ sh:class schema:URL ;
 sh:path prov:wasDerivedFrom ], [ sh:class schema:URL ;
 sh:path schema:sameAs ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("0", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path renku:originalIdentifier ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path renku:slug ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:identifier ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:name ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:moreThanOrEquals schema:dateCreated ;
 sh:nodeKind sh:Literal ;
 sh:path prov:invalidatedAtTime ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:dateCreated ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:datePublished ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:description ], [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ;
 sh:path schema:keywords ], [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ;
 sh:path schema:url ], [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ;
 sh:path schema:version ], [ sh:nodeKind sh:Literal ;
 sh:or ( [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ] [ sh:nodeKind sh:BlankNodeOrIRI ] ) ;
 sh:path schema:license ] ;
 sh:targetClass schema:Dataset ];
 Focus Node: <https://david.dev.renku.ch/datasets/617f2eaf232917bb0b046893e3a043e7>;
 Value Node: <https://david.dev.renku.ch/datasets/652eb758-d7d4-4a7f-b76e-6157c21adb20/images/0>;
 Result Path: schema:image;
 Message: Node <https://david.dev.renku.ch/datasets/617f2eaf232917bb0b046893e3a043e7> is closed. It cannot have value: <https://david.dev.renku.ch/datasets/652eb758-d7d4-4a7f-b76e-6157c21adb20/images/0>;
 Constraint Violation in ClosedConstraintComponent (http://www.w3.org/ns/shacl#ClosedConstraintComponent):;
 Severity: sh:Violation;
 Source Shape: [ rdf:type sh:NodeShape ;
 sh:closed Literal("true" = True, datatype=xsd:boolean) ;
 sh:ignoredProperties ( rdf:type ) ;
 sh:property [ sh:class schema:DigitalDocument ;
 sh:path schema:hasPart ], [ sh:class schema:Language ;
 sh:path schema:inLanguage ], [ sh:class schema:Person ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:path schema:creator ], [ sh:class schema:PublicationEvent ;
 sh:path schema:subjectOf ], [ sh:class schema:URL ;
 sh:path prov:wasDerivedFrom ], [ sh:class schema:URL ;
 sh:path schema:sameAs ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("0", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path renku:originalIdentifier ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path renku:slug ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:identifier ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:minCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:name ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:moreThanOrEquals schema:dateCreated ;
 sh:nodeKind sh:Literal ;
 sh:path prov:invalidatedAtTime ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:dateCreated ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:datePublished ], [ sh:datatype xsd:string ;
 sh:maxCount Literal("1", datatype=xsd:integer) ;
 sh:nodeKind sh:Literal ;
 sh:path schema:description ], [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ;
 sh:path schema:keywords ], [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ;
 sh:path schema:url ], [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ;
 sh:path schema:version ], [ sh:nodeKind sh:Literal ;
 sh:or ( [ sh:datatype xsd:string ;
 sh:nodeKind sh:Literal ] [ sh:nodeKind sh:BlankNodeOrIRI ] ) ;
 sh:path schema:license ] ;
 sh:targetClass schema:Dataset ];
 Focus Node: <https://david.dev.renku.ch/datasets/652eb758d7d44a7fb76e6157c21adb20>;
 Value Node: <https://david.dev.renku.ch/datasets/652eb758-d7d4-4a7f-b76e-6157c21adb20/images/0>;
 Result Path: schema:image;
 Message: Node <https://david.dev.renku.ch/datasets/652eb758d7d44a7fb76e6157c21adb20> is closed. It cannot have value: <https://david.dev.renku.ch/datasets/652eb758-d7d4-4a7f-b76e-6157c21adb20/images/0>;
 ;
 Couldn't export: Invalid Knowledge Graph data;
```



#13 Probably due to data/*/metadata.yml being stored in git lfs, verify and ignore if that's the case
```
aec9988f217a1022a16f4410d9f55bcf585c68d7 |       2493 | luis.salamanca/democ_sentiment

Applying migration m_0003__0_pyld2...;
Applying migration m_0003__1_jsonld...;
Applying migration m_0003__2_initial...;
Error: Couldn't execute migration;
;
Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/core/management/migrate.py", line 157, in migrate;
module.migrate(migration_context);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0003__2_initial.py", line 45, in migrate;
_migrate_datasets_pre_v0_3(client);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/m_0003__2_initial.py", line 87, in _migrate_datasets_pre_v0_3;
dataset = Dataset.from_yaml(old_path, client);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/models/v3.py", line 175, in from_yaml;
self = DatasetSchemaV3(client=client, commit=commit).load(data);
File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 719, in load;
return self._do_load(;
File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 844, in _do_load;
processed_data = self._invoke_load_processors(;
File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 1090, in _invoke_load_processors;
data = self._invoke_processors(;
File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 1220, in _invoke_processors;
data = processor(data, many=many, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/models/v3.py", line 360, in fix_files_context;
data = migrate_types(data);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/utils/__init__.py", line 169, in migrate_types;
replace_types(data);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrations/utils/__init__.py", line 152, in replace_types;
for key, value in data_.items():;
AttributeError: 'str' object has no attribute 'items';
;
The above exception was the direct cause of the following exception:;
;
Traceback (most recent call last):;
File "/usr/lib/python3.8/site-packages/renku/cli/exception_handler.py", line 89, in main;
return super().main(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/click/core.py", line 1062, in main;
rv = self.invoke(ctx);
File "/usr/lib/python3.8/site-packages/click/core.py", line 1668, in invoke;
return _process_result(sub_ctx.command.invoke(sub_ctx));
File "/usr/lib/python3.8/site-packages/click/core.py", line 1404, in invoke;
return ctx.invoke(self.callback, **ctx.params);
File "/usr/lib/python3.8/site-packages/click/core.py", line 763, in invoke;
return __callback(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/cli/migrate.py", line 109, in migrate;
result = command.build().execute(;
File "/usr/lib/python3.8/site-packages/renku/core/management/command_builder/command.py", line 256, in execute;
hook(self, context, result, *args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/command_builder/command.py", line 195, in _post_hook;
raise result.error;
File "/usr/lib/python3.8/site-packages/renku/core/management/command_builder/command.py", line 242, in execute;
output = context["click_context"].invoke(self._operation, *args, **kwargs);
File "/usr/lib/python3.8/site-packages/click/core.py", line 763, in invoke;
return __callback(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/commands/migrate.py", line 164, in _migrate_project;
return migrate(;
File "/usr/lib/python3.8/site-packages/inject/__init__.py", line 342, in injection_wrapper;
return sync_func(*args, **kwargs);
File "/usr/lib/python3.8/site-packages/renku/core/management/migrate.py", line 159, in migrate;
raise MigrationError("Couldn't execute migration") from e;
renku.core.errors.MigrationError: Couldn't execute migration;
```


We can fix all these in one PR.

--- a/renku/core/management/migrate.py
+++ b/renku/core/management/migrate.py
@@ -57,6 +57,7 @@ from renku.core.management.migrations.utils import (
     is_using_temporary_datasets_path,
     read_project_version,
 )
+from renku.core.management.workflow.plan_factory import RENKU_TMP
 from renku.core.utils import communication
 from renku.core.utils.git import is_valid_git_repository
 
@@ -158,14 +159,26 @@ def migrate(
             except (Exception, BaseException) as e:
                 raise MigrationError("Couldn't execute migration") from e
             n_migrations_executed += 1
-    if n_migrations_executed > 0 and not is_using_temporary_datasets_path():
-        client._project = None  # NOTE: force reloading of project metadata
-        client.project.version = str(version)
-        project_gateway.update_project(client.project)
+    if not is_using_temporary_datasets_path():
+        if n_migrations_executed > 0:
+            client._project = None  # NOTE: force reloading of project metadata
+            client.project.version = str(version)
+            project_gateway.update_project(client.project)
 
-        communication.echo(f"Successfully applied {n_migrations_executed} migrations.")
+            communication.echo(f"Successfully applied {n_migrations_executed} migrations.")
+
+        _remove_untracked_renku_files(renku_path=client.renku_path)
 
     return n_migrations_executed != 0, template_updated, docker_updated
+
+
+def _remove_untracked_renku_files(renku_path):
+    from renku.core.management.datasets import DatasetsApiMixin
+
+    untracked_paths = [RENKU_TMP, DatasetsApiMixin.CACHE, "vendors"]
+    for path in untracked_paths:
+        path = renku_path / path
+        shutil.rmtree(path, ignore_errors=True)
 
 
 @inject.autoparams()

--- a/renku/core/management/migrations/utils/conversion.py
+++ b/renku/core/management/migrations/utils/conversion.py
@@ -153,6 +153,20 @@ def convert_dataset(dataset: old_datasets.Dataset, client, revision: str) -> Tup
 
         return Url(url_id=Dataset.generate_id(identifier=Path(path).name))
 
+    def convert_license(license):
+        if not license:
+            return license
+        elif isinstance(license, (Url, str)):
+            return license
+        elif isinstance(license, dict) and len(license) == 1:
+            return list(license.values())[0]
+        elif isinstance(license, dict) and "@id" in license:
+            return license["@id"]
+        elif isinstance(license, list) and len(license) == 1:
+            return license[0]
+
+        return str(license)
+
     tags = [_convert_dataset_tag(tag) for tag in (dataset.tags or [])]
 
     return (
@@ -169,7 +183,7 @@ def convert_dataset(dataset: old_datasets.Dataset, client, revision: str) -> Tup
             images=[_convert_image_object(image) for image in (dataset.images or [])],
             in_language=_convert_language(dataset.in_language),
             keywords=dataset.keywords,
-            license=dataset.license,
+            license=convert_license(dataset.license),
             name=dataset.name,
             project_id=client.project.id,
             initial_identifier=_convert_dataset_identifier(dataset.initial_identifier),

--- a/renku/data/shacl_shape.json
+++ b/renku/data/shacl_shape.json
@@ -401,6 +401,12 @@
                }
             },
             {
+               "path": "schema:image",
+               "sh:class": {
+                  "@id": "schema:ImageObject"
+               }
+            },
+            {
                "nodeKind": "sh:Literal",
                "path": "schema:keywords",
                "datatype": {
@@ -501,6 +507,35 @@
                      "nodeKind": "sh:IRI"
                   }
                ],
+               "maxCount": 1
+            }
+         ]
+      },
+      {
+         "@id": "_:ImageObjectShape",
+         "@type": "sh:NodeShape",
+         "ignoredProperties": [
+            {
+               "@id": "rdf:type"
+            }
+         ],
+         "closed": true,
+         "targetClass": "schema:ImageObject",
+         "property": [
+            {
+               "path": "schema:contentUrl",
+               "nodeKind": "sh:Literal",
+               "datatype": {
+                  "@id": "xsd:string"
+               },
+               "maxCount": 1
+            },
+            {
+               "path": "schema:position",
+               "nodeKind": "sh:Literal",
+               "datatype": {
+                  "@id": "xsd:integer"
+               },
                "maxCount": 1
             }
          ]
@@ -770,9 +805,18 @@
             },
             {
                "path": "prov:agent",
-               "sh:class": {
-                  "@id": "prov:SoftwareAgent"
-               },
+               "or": [
+                  {
+                     "sh:class": {
+                        "@id": "prov:SoftwareAgent"
+                     }
+                  },
+                  {
+                     "sh:class": {
+                        "@id": "schema:Person"
+                     }
+                  }
+               ],
                "minCount": 1,
                "maxCount": 1
             }

--- a/tests/cli/test_migrate.py
+++ b/tests/cli/test_migrate.py
@@ -24,9 +24,11 @@ from pathlib import Path
 import pytest
 
 from renku.cli import cli
+from renku.core.management import RENKU_HOME
 from renku.core.management.client import LocalClient
 from renku.core.management.dataset.datasets_provenance import DatasetsProvenance
 from renku.core.management.migrate import SUPPORTED_PROJECT_VERSION, get_migrations
+from renku.core.management.workflow.plan_factory import RENKU_TMP
 from renku.core.metadata.repository import Repository
 from renku.core.models.dataset import RemoteEntity
 from tests.utils import format_result_exception
@@ -164,6 +166,11 @@ def test_migrations_runs(isolated_runner, old_project):
     result = isolated_runner.invoke(cli, ["migrate"])
     assert 0 == result.exit_code, format_result_exception(result)
     assert "No migrations required." in result.output
+
+    tmp_path = os.path.join(RENKU_HOME, RENKU_TMP)
+    paths = [c.b_path for c in old_project.head.commit.get_changes() if tmp_path in c.b_path]
+
+    assert 0 == len(paths), ", ".join(paths)
 
 
 @pytest.mark.migration

